procrastinate/aiopg_connector.py

```diff
@@ -103,31 +56,86 @@ async def create_with_pool_async(
+            the function used by psycopg2. See the `psycopg2 doc`_.
+        json_loads:
+            The JSON loads function to use for deserializing job arguments. Defaults
+            to the function used by psycopg2. See the `psycopg2 doc`_. Unused if pool
+            is passed.
+        """
+        self._pool: Optional[aiopg.Pool] = None
+        self.json_dumps = json_dumps
+        self.json_loads = json_loads
+        self._pool_args = self._adapt_pool_args(kwargs, json_loads)
+        self._lock = asyncio.Lock()
+
+    @staticmethod
+    def _adapt_pool_args(
+        pool_args: Dict[str, Any], json_loads: Optional[Callable]
+    ) -> Dict[str, Any]:
         """
-        base_on_connect = kwargs.pop("on_connect", None)
+        Adapt the pool args for ``aiopg``, using sensible defaults for Procrastinate.
+        """
+        base_on_connect = pool_args.pop("on_connect", None)
 
         async def on_connect(connection):
             if base_on_connect:
                 await base_on_connect(connection)
             if json_loads:
                 psycopg2.extras.register_default_jsonb(connection.raw, loads=json_loads)
 
-        defaults = {
+        final_args = {
             "dsn": "",
             "enable_json": False,
             "enable_hstore": False,
             "enable_uuid": False,
             "on_connect": on_connect,
             "cursor_factory": RealDictCursor,
         }
-        if "maxsize" in kwargs:
-            kwargs["maxsize"] = max(2, kwargs["maxsize"])
+        if "maxsize" in pool_args:
+            pool_args["maxsize"] = max(2, pool_args["maxsize"])
+
+        final_args.update(pool_args)
+        return final_args
+
+    async def close_async(self) -> None:
+        """
+        Close the pool and awaits all connections to be released.
+        """
+        if self._pool:
+            self._pool.close()
+            await self._pool.wait_closed()
+            self._pool = None
+
+    def _wrap_json(self, arguments: Dict[str, Any]):
+        return {
+            key: Json(value, dumps=self.json_dumps)
+            if isinstance(value, dict)
+            else value
+            for key, value in arguments.items()
+        }
 
-        defaults.update(kwargs)
+    @staticmethod
+    async def _create_pool(pool_args: Dict[str, Any]) -> aiopg.Pool:
+        return await aiopg.create_pool(**pool_args)
 
-        pool = await aiopg.create_pool(**defaults)
+    def set_pool(self, pool: aiopg.Pool) -> None:
+        """
+        Set the connection pool. Raises an exception if the pool is already set.
+        """
+        if self._pool:
+            raise exceptions.PoolAlreadySet
+        self._pool = pool
 
-        return cls(pool=pool, json_dumps=json_dumps, json_loads=json_loads)
+    async def _get_pool(self) -> aiopg.Pool:
+        async with self._lock:
+            if not self._pool:
+                self.set_pool(await self._create_pool(self._pool_args))
+        return self._pool
```

Comment:
I think it doesn't hurt to add a fast path to avoid taking the lock every single time we need the pool:
```python
if self._pool:
    return self._pool

async with self._lock:
    ...
```

Reply:
Yep. We added the fast path, then got confused and thought it wouldn't work. But you're right, that works.

Reply:
It works only if you re-test the value of self._pool inside the lock. We were confused because we tried moving the if instead of duplicating it.

Reply:
An alternative would be to put the lock in the `_create_pool` function. That would allow limiting logic & IO being mixed up !

Reply:
Ah no forget it !

Reply:
Yep.